From a8d4f9054ac845041e00dbdb89b184f20091b6f0 Mon Sep 17 00:00:00 2001
From: Yuriy Umanets <yuriy_umanets@xyratex.com>
Date: Mon, 30 May 2011 03:59:31 -0600
Subject: [PATCH 098/158] - fixed bug in c2_md_rename_fop_state() with checking
 version of cob and sbody; - fixed bug in c2_md_rename_fop_state(). After
 rename we need to update cob version to new one from sbody; - changed order
 of args in store_rename;

---
 mdservice/md_foms.c | 39 +++++++++++++++++++++++++--------------
 mdstore/mdstore.c   |  2 +-
 mdstore/mdstore.h   |  2 +-
 3 files changed, 27 insertions(+), 16 deletions(-)

diff --git a/mdservice/md_foms.c b/mdservice/md_foms.c
index 5297d6d..c75b9f8 100644
--- a/mdservice/md_foms.c
+++ b/mdservice/md_foms.c
@@ -406,6 +406,7 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         struct c2_fid             tfid;
         struct c2_fid             pfid_tgt;
         struct c2_fid             pfid_src;
+        struct c2_cob_attr        attr;
         struct c2_service        *svc;
         int                       rc;
 
@@ -428,6 +429,7 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         sbody = &req->r_sbody;
         tbody = &req->r_tbody;
 
+        c2_md_fop_cob2attr(&attr, sbody);
         c2_md_make_fid(&pfid_src, &sbody->b_pfid);
         c2_md_make_fid(&pfid_tgt, &tbody->b_pfid);
         c2_md_make_fid(&tfid, &sbody->b_tfid);
@@ -438,30 +440,39 @@ static int c2_md_rename_fom_state(struct c2_fom *fom)
         if (rc)
                 goto out;
 
-        if (cob->co_nsrec.cnr_version > sbody->b_version) {
-                struct c2_bitstring *src;
-                struct c2_bitstring *tgt;
+        if (sbody->b_version > cob->co_nsrec.cnr_version) {
+                struct c2_bitstring *name_src;
+                struct c2_bitstring *name_tgt;
                 
-                src = c2_bitstring_alloc(req->r_sname.n_name, 
-                                         req->r_sname.n_len);
-                if (!src) {
+                name_src = c2_bitstring_alloc(req->r_sname.n_name, 
+                                              req->r_sname.n_len);
+                if (!name_src) {
                         rc = -ENOMEM;
                         goto out;
                 }
                 
-                tgt = c2_bitstring_alloc(req->r_tname.n_name, 
-                                         req->r_tname.n_len);
-                if (!tgt) {
-                        c2_bitstring_free(src);
+                name_tgt = c2_bitstring_alloc(req->r_tname.n_name, 
+                                              req->r_tname.n_len);
+                if (!name_tgt) {
+                        c2_bitstring_free(name_src);
                         rc = -ENOMEM;
                         goto out;
                 }
 
                 rc = c2_md_store_rename(site->s_mdstore, &pfid_tgt, 
-                                        &pfid_src, cob, src, tgt,
-                                        &ctx->fc_tx->tx_dbtx);
-                c2_bitstring_free(src);
-                c2_bitstring_free(tgt);
+                                        &pfid_src, cob, name_tgt, 
+                                        name_src, &ctx->fc_tx->tx_dbtx);
+                c2_bitstring_free(name_src);
+                c2_bitstring_free(name_tgt);
+
+                /*
+                 * Let's update attributes given we found that sbody contains
+                 * new set of then. Most of all we need to update version.
+                 */
+                if (rc == 0) {
+                        rc = c2_md_store_setattr(site->s_mdstore, cob,
+                                                 &attr, &ctx->fc_tx->tx_dbtx);
+                }
         }
         c2_cob_put(cob);
         if (rc == 0) {
diff --git a/mdstore/mdstore.c b/mdstore/mdstore.c
index 754c257..4b07f9d 100644
--- a/mdstore/mdstore.c
+++ b/mdstore/mdstore.c
@@ -340,8 +340,8 @@ int c2_md_store_rename(struct c2_md_store       *md,
                        struct c2_fid            *pfid_tgt,
                        struct c2_fid            *pfid_src,
                        struct c2_cob            *cob,
-                       struct c2_bitstring      *name_src,
                        struct c2_bitstring      *name_tgt,
+                       struct c2_bitstring      *name_src,
                        struct c2_db_tx          *tx)
 {
         struct c2_cob_nskey  *srckey;
diff --git a/mdstore/mdstore.h b/mdstore/mdstore.h
index 7940ed0..5a17fa1 100644
--- a/mdstore/mdstore.h
+++ b/mdstore/mdstore.h
@@ -225,8 +225,8 @@ int c2_md_store_rename(struct c2_md_store *md,
                        struct c2_fid *pfid_tgt,
                        struct c2_fid *pfid_src,
                        struct c2_cob *cob,
-                       struct c2_bitstring *name_src,
                        struct c2_bitstring *name_tgt,
+                       struct c2_bitstring *name_src,
                        struct c2_db_tx *tx);
 
 /**
-- 
1.8.3.2

